cirq-core/cirq/transformers/analytical_decompositions/quantum_shannon_decomposition.py

```diff
@@ -137,10 +198,30 @@ def quantum_shannon_decomposition(
     # Yield ops from multiplexed Ry part
     yield from _multiplexed_cossin(qubits, theta, ops.ry)
 
+    # Optimization A.1 in Shende et al. - the last CZ gate in the multiplexed Ry part
+    # is merged into the generic multiplexor (u1, u2)
+    # This gate is CZ(qubits[1], qubits[0]) = CZ(qubits[0], qubits[1])
+    # as CZ is symmetric.
+    # For the u1⊕u2 multiplexor operator:
+    # as u1 is the operator in case qubits[0] = |0>,
+    # and u2 is the operator in case qubits[0] = |1>
+    # we can represent the merge by phasing u2 with Z ⊗ I
+    cz_diag = np.concatenate((np.ones(n >> 2), np.full(n >> 2, -1)))
+    u2 = u2 @ np.diag(cz_diag)
+
     # Yield ops from decomposition of multiplexed u1/u2 part
     yield from _msb_demuxer(qubits, u1, u2)
 
 
+def _global_phase_difference(u: np.ndarray, ops: list[cirq.Operation]) -> float:
```

Comment:
do we still need this? I thought that the analytical decompositions now preserve global phase https://github.com/quantumlib/Cirq/issues/6523, is this not the case?

Reply:
I tried to remove it, but unit tests are failing. Looking at #6523 , it seems to me it was mistakenly closed by #7118 . #7118 changes the MatrixGate decomposition, but #6523 requires updating `two_qubit_to_cz`, a different thing.